pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -1093,6 +1123,77 @@ func (nc *Controller) handleDisruption(zoneToNodeConditions map[string][]*v1.Nod
 	}
 }
 
+func (nc *Controller) podUpdated(oldPod, newPod *v1.Pod) {
+	if len(newPod.Spec.NodeName) != 0 && (oldPod == nil || newPod.Spec.NodeName != oldPod.Spec.NodeName) {
+		nc.podUpdateQueue.Add(newPod)
+	}
+}
+
+func (nc *Controller) doPodProcessingWorker() {
+	for {
+		obj, shutdown := nc.podUpdateQueue.Get()
+		// "nodeUpdateQueue" will be shutdown when "stopCh" closed;
+		// we do not need to re-check "stopCh" again.
+		if shutdown {
+			return
+		}
+
+		pod := obj.(*v1.Pod)
+		nodeName := pod.Spec.NodeName
+
+		func() {
+			nc.nodeHealthLock.Lock()
+			defer nc.nodeHealthLock.Unlock()
+			defer nc.podUpdateQueue.Done(pod)
+
+			status, exists := nc.nodeHealthMap[nodeName]
+			if !exists {
+				return
+			}
+
+			node, err := nc.nodeLister.Get(nodeName)
+			if err != nil {
+				klog.Errorf("Failed to list node %v.", nodeName)
+				return
+			}
+			decisionTimestamp := nc.now()
+			if status.currentReadyCondition != nil {
+				if nc.useTaintBasedEvictions {
+					if status.observedReadyCondition.Status == v1.ConditionFalse &&
+						decisionTimestamp.After(status.readyTransitionTimestamp.Add(nc.podEvictionTimeout)) {
+						if nc.evictPods(node) {
+							klog.V(2).Infof("Node is NotReady. Adding Pods on Node %s to eviction queue: %v is later than %v + %v",
+								nodeName,
+								decisionTimestamp,
+								status.readyTransitionTimestamp,
+								nc.podEvictionTimeout,
+							)
+						}
+					}
+					if status.observedReadyCondition.Status == v1.ConditionUnknown &&
+						decisionTimestamp.After(status.probeTimestamp.Add(nc.podEvictionTimeout)) {
+						if nc.evictPods(node) {
+							klog.V(2).Infof("Node is unresponsive. Adding Pods on Node %s to eviction queues: %v is later than %v + %v",
+								nodeName,
+								decisionTimestamp,
+								status.readyTransitionTimestamp,
+								nc.podEvictionTimeout-status.gracePeriod,
+							)
+						}
+					}
+				}
+
+				if status.currentReadyCondition.Status != v1.ConditionTrue &&
```

Comment:
I think this is incorrect (correct me if I'm wrong).

currentReadyCondition.Status != observedReadyCondition.Status only if in the last loop of monitorNodeHealth the node has changed the status.

What if e.g.
* in first loop monitorNodeHealth node goes unhealthy, so currentReadyCondition.Status != observedReadyCondition.Status
* in the second loop monitorNodeHealth node is still unhealthy so currentReadyCondition.Status == observedReadyCondition.Status
* after that podUpdate appears and you don't enter that if (but I think you should).

Reply:
Ok. I think I understand the idea now.
[This if](https://github.com/kubernetes/kubernetes/blob/a268eefc31e3a5d13ea867d47c1a4b060e52343d/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L818) checks only the transition so marking pods is only done once. However, when pod is added I need to check against the current state.
Good catch.